zebedee-cms/src/main/java/com/github/onsdigital/zebedee/keyring/store/CollectionKeyStoreImpl.java

```diff
@@ -0,0 +1,193 @@
+    static final String COLLECTION_KEY_NOT_FOUND_ERR = "collectionKey not found";
+    static final String KEY_DECRYPTION_ERR = "error while decrypting collectionKey file";
+    static final String WRITE_KEY_ERR = "error writing collection key to store";
+    static final String COLLECTION_KEY_ALREADY_EXISTS_ERR = "collectionKey for this collection ID already exists";
+    static final String ENCRYPTION_ALGORITHM = "AES";
+    static final String CIPHER_ALGORITHM = "AES/CBC/PKCS5Padding";
+    static final String FILE_ACCESS_MODE = "rw";
+
+    private Path keyringDir;
+    private transient SecretKey masterKey;
+    private transient IvParameterSpec masterIv;
+
+    /**
+     * <p>
+     * Create a new CollectionKeyStore instance.
+     * </p>
+     *
+     * <p>
+     * <b>Important:</b> It is strongly recommended that the {@link SecretKey} & {@link IvParameterSpec}
+     * are not modified once they have been used to store collection keys. Changing these inputs will result in this
+     * object no longer being able to decrypt collection key files written with the previous encryption key.
+     * <p>
+     * Reverting the key/init vector values will resolve this issue for keys written with the original encryption
+     * parameters however, the same issue will now be the case for any key files written using the modified encryption
+     * params.
+     * </p>
+     *
+     * <p>
+     * <b>
+     * If a encryption parameters change is necessary it is strongly recommended you migrate all existing keys to
+     * use the updated encryption values.</b>
+     * </p>
+     *
+     * @param keyringDir the {@link Path} to read the collection encryption keys files from.
+     * @param masterKey  the {@link SecretKey} to use when decrypting the collection key files.
+     * @param masterIv   the {@link IvParameterSpec} to use when initializing the encryption {@link Cipher}.
+     */
+    public CollectionKeyStoreImpl(final Path keyringDir, final SecretKey masterKey, final IvParameterSpec masterIv) {
+        this.keyringDir = keyringDir;
+        this.masterKey = masterKey;
+        this.masterIv = masterIv;
+    }
+
+    @Override
+    public SecretKey read(final String collectionID) throws KeyringException {
+        validateRead(collectionID);
+
+        byte[] keyBytes = null;
+        try (
+                RandomAccessFile raf = new RandomAccessFile(getKeyPath(collectionID), FILE_ACCESS_MODE);
+                FileChannel channel = raf.getChannel();
+                InputStream in = Channels.newInputStream(channel);
+                CipherInputStream cin = new CipherInputStream(in, getDecryptCipher());
+        ) {
+            // FileLock documentation states the lock will released when the FileChannel used to acquire it is closed.
+            // FileChannel implements autoclosable and in this case is defined inside a try-with-resources which
+            // guarantees auto closeable resourses are closed.
+            channel.lock();
```

Comment:
hmmm. My understanding after reading that documentation is that a FileLock is actually not suitable mechanism for controlling multi thread access to a file. Good spot.

I was hoping to avoid this but I think the only way implement this functionality with files on disk is to use a lock object on for the entire class. I've had to do this elsewhere to fix race condition bugs, it's not nice and it kills performance but I think that is the cost of data integrity when your "database" is files on disk.

(For the record I'm not responsible for that design decision - it's been a long running bone of contention for me).